### PR TITLE
Enable extensions with the Django integration

### PR DESCRIPTION
**Pull request by whiskybar**
I thought Ariadne could support Apollo tracing in a Django GraphQL project: we need to pass `extensions` down to `graphql_sync`, it's as easy as this PR.

### CODE REVIEW

**Comment by whiskybar** [comment]
Hi there, is there something more I should do for this code to be accepted?

**Comment by rafalp** [comment]
Hello, sorry for not giving this attention for so long!

In long run we would like to extract `ariadne.contrib.django` into separate package (`ariadne-django`), but in the meantime I think only part missing here would be tests asserting that extension is passed by the view to Ariadne's query executor. When this lands, I'll be happy to merge and release next Ariadne with this feature.

**Comment by tswicegood** [comment]
FWIW, I've run this locally and it works as expected.

**Comment by markedwards** [comment]
I'd like to see this merged.  @whiskybar is there anything I can do to help move this along?

**Comment by whiskybar** [comment]
Getting back to you. I finally added the tests. I should hope the PR is ready? Sorry for my delay.

File: ariadne/contrib/django/views.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
     GraphQLResult,
     RootValue,
     ValidationRules,
+    Extension,
```
Comment (rafalp) [inline ariadne/contrib/django/views.py:24]: Please sort imports alphabetically.

File: ariadne/contrib/django/views.py [hunk 2]
```diff
@@ -38,6 +39,7 @@ class GraphQLView(TemplateView):
     logger = None
     validation_rules: Optional[ValidationRules] = None
     error_formatter: Optional[ErrorFormatter] = None
+    extensions: Optional[List[Type[Extension]]] = None
```
Comment (rafalp) [inline ariadne/contrib/django/views.py:42]: This type doesn't match one on `wsgi.GraphQL` - extensions can be callable:

```
Extensions = Union[
    Callable[[Any, Optional[ContextValue]], ExtensionList], ExtensionList
]
```

```
extensions: Optional[Extensions] = None
```

File: ariadne/contrib/django/views.py [hunk 3]
```diff
@@ -120,5 +122,6 @@ def execute_query(self, request: HttpRequest, data: dict) -> GraphQLResult:
             debug=settings.DEBUG,
             logger=self.logger,
             error_formatter=self.error_formatter or format_error,
+            extensions=self.extensions,
```
Comment (rafalp) [inline ariadne/contrib/django/views.py:125]: There will need to be a logic testing for extensions being callable:

```
if callable(self.extensions):
    extensions = self.extensions(request, context_value)
else:
    extensions = self.extensions
```

**Comment by rafalp** [comment]
@whiskybar test looks good, but I've only just realized that we support callable `extensions` to let server authors programmatically set extensions based on request and context. I've included code examples but could you please update view to follow `get_context_for_request` and `get_extensions_for_request` implementation as done in `ariadne.wsgi.GraphQL` class?

Thank you and sorry about being so late with proper review!

**Comment by whiskybar** [comment]
Meh, this will need some care with typing as GraphQL descends from TemplateView. More on that later.

**Comment by whiskybar** [comment]
@rafalp can you take another look at the PR now?

File: ariadne/contrib/django/views.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import json
-from typing import Optional, cast
+from typing import Any, Callable, cast, List, Optional, Type, Union
```
Comment (rafalp) [inline ariadne/contrib/django/views.py:2]:
```suggestion
from typing import Any, Callable, List, Optional, Type, Union, cast
```

File: tests/django/test_configuration.py [hunk 5]
```diff
@@ -4,6 +4,7 @@
 import pytest
 from django.test import override_settings
 
+from ariadne.types import ExtensionSync
```
Comment (rafalp) [inline tests/django/test_configuration.py:7]: `.types` should be imported after `.contrib`

File: tests/django/test_configuration.py [hunk 6]
```diff
@@ -131,6 +132,36 @@ def test_custom_validation_rules_function_is_called_with_context_value(
     get_validation_rules.assert_called_once_with({"test": "TEST-CONTEXT"}, ANY, ANY)
 
 
+def test_extensions_are_called_on_query_executor(request_factory, schema):
```
Comment (rafalp) [inline tests/django/test_configuration.py:135]:
```suggestion
def test_extensions_function_result_is_passed_to_query_executor(
    request_factory, schema
):
```

File: tests/django/test_configuration.py [hunk 7]
```diff
@@ -131,6 +132,36 @@ def test_custom_validation_rules_function_is_called_with_context_value(
...
+        {"query": "{ status }"},
+        context_value=context,
+        extensions=get_extensions,
+    )
+    get_extensions.assert_called_once()
+    extension.request_started.assert_called_once_with(context)
+    extension.request_finished.assert_called_once_with(context)
+
+
+def test_extension_is_passed_to_query_executor(request_factory, schema):
```
Comment (rafalp) [inline tests/django/test_configuration.py:151]:
```suggestion
def test_extension_from_option_is_passed_to_query_executor(
    request_factory, schema
):
```

**Comment by rafalp** [comment]
Looks like last batch of things to change.

**Comment by rafalp** [comment]
All-right, lets ship this in Ariadne 0.12!

**Pull request merged**